PWX-32313 : Automate tests for FADA raw block volume feature

ashinde-px opened the pull request:
<!--
  Make sure to have done the following:
  [] Signed off your work as per the DCO.
  [] Add unit-tests
-->

**What this PR does / why we need it**:
Automate tests for FADA raw block volume feature

**Which issue(s) this PR fixes** (optional)
Closes #PWX-32313

**Special notes for your reviewer**:

Rohit-PX commented on drivers/volume/portworx/portworx.go:
```diff
@@ -222,6 +227,16 @@ type statusJSON struct {
 	Cmd    string
 }
 
+func (d *portworx) executeCommandToPod(podCmd, podName, podNamespace string) (string, error) {
```
This is a redundant method, there is an existing method in node driver to execute command inside pods, example of it's usage:
https://github.com/portworx/torpedo/blob/99b2904b4de9eacc906a3439084d8eb880c015a6/tests/basic/sharedv4_svc_test.go#L1713

ashinde-px replied:
done.

jainsmit commented on drivers/volume/portworx/portworx.go:
```diff
@@ -104,6 +105,10 @@ const (
 	pxServiceLocalEndpoint                    = "portworx-service.kube-system.svc.cluster.local"
 	mountGrepVolume                           = "mount | grep %s"
 	mountGrepFirstColumn                      = "mount | grep %s | awk '{print $1}'"
+	ddCopy                                    = "dd if=%s of=%s bs=${%s} count=1 seek=0"
```
Can we use these commands as part of application spec using raw block ?

ashinde-px replied:
Done.

jainsmit commented on drivers/volume/portworx/portworx.go:
```diff
@@ -1623,6 +1638,50 @@ func (d *portworx) ValidatePureFaCreateOptions(volumeName string, FStype string,
 }
 
+// ValidatePureRawBlockVolumes validates rawblock pure volumes
+func (d *portworx) ValidatePureRawBlockVolumes(volumeName string, pvcObj *corev1.PersistentVolumeClaim) error {
```
This method is not needed if all these commands are part of application spec itself and either ValidateContextForPureVolumesSDK or ValidateContext can be use to validate the application.

ashinde-px replied:
Done.

vinayakshnd reviewed (approved):
LGTM

jainsmit commented on drivers/scheduler/k8s/k8s.go:
```diff
@@ -4140,16 +4145,19 @@ func (k *K8s) ResizeVolume(ctx *scheduler.Context, configMapName string) ([]*vol
-			}
-			if shouldResize {
-				vol, err := k.ResizePVC(ctx, updatedPVC, 1)
+			// For raw block volumes resize is failing hence skipping test for it. defect filed - PWX-32793
```
Actually instead of disabling here please disable the test for raw block and make sure to enable again after bug got fixed.

jainsmit commented on drivers/scheduler/k8s/k8s.go:
```diff
@@ -4140,16 +4145,19 @@ func (k *K8s) ResizeVolume(ctx *scheduler.Context, configMapName string) ([]*vol
-			if shouldResize {
-				vol, err := k.ResizePVC(ctx, updatedPVC, 1)
+			// For raw block volumes resize is failing hence skipping test for it. defect filed - PWX-32793
+			if *updatedPVC.Spec.VolumeMode != corev1.PersistentVolumeBlock {
```
Actually instead of disabling here please disable the test for raw block and make sure to enable again after bug got fixed.

vinayakshnd replied:
I would suggest to keep a TODO note here with Bug ID that once the Bug is fixed, enable below test.

ashinde-px replied:
Yes actually we are skeeping the test at context level only instead of application actions. did changes as suggested and resolving the comment now.

jainsmit commented on drivers/scheduler/k8s/k8s.go:
```diff
@@ -4169,16 +4177,19 @@ func (k *K8s) ResizeVolume(ctx *scheduler.Context, configMapName string) ([]*vol
-				if shouldResize {
-					vol, err := k.ResizePVC(ctx, &pvc, 1)
+				// For raw block volumes resize is failing hence skipping test for it
+				if *pvc.Spec.VolumeMode != corev1.PersistentVolumeBlock {
```
Same..

ashinde-px replied:
Done

jainsmit commented on drivers/scheduler/k8s/k8s.go:
```diff
@@ -4199,16 +4210,19 @@ func (k *K8s) ResizeVolume(ctx *scheduler.Context, configMapName string) ([]*vol
-				if shouldResize {
-					vol, err := k.ResizePVC(ctx, &pvc, 1)
+				// For raw block volumes resize is failing hence skipping test for it
+				if *pvc.Spec.VolumeMode != corev1.PersistentVolumeBlock {
```
Same

ashinde-px replied:
Done.

jainsmit commented on drivers/volume/portworx/portworx.go:
```diff
@@ -1287,7 +1290,7 @@ func (d *portworx) ValidateCreateVolume(volumeName string, params map[string]str
 
 	// Validate Device Path for a Volume
-	if vol.Spec.ProxySpec != nil && vol.Spec.ProxySpec.ProxyProtocol == api.ProxyProtocol_PROXY_PROTOCOL_PURE_BLOCK {
+	if vol.Spec.ProxySpec != nil && vol.Spec.ProxySpec.ProxyProtocol == api.ProxyProtocol_PROXY_PROTOCOL_PURE_BLOCK && !isRaw {
```
Why we are not checking device path for raw block volume ? it will be a direct access volume right ..

ashinde-px replied:
Done.

jainsmit commented on tests/common.go:
```diff
@@ -1143,27 +1144,31 @@ func ValidatePureVolumeStatisticsDynamicUpdate(ctx *scheduler.Context, errChan .
+			processError(err, errChan...)
+			fmt.Println("sleeping to let volume usage get reflected")
+			// wait until the backends size is reflected before making the REST call
+			time.Sleep(time.Minute * 2)
```
Is there any other way to make sure size is reflecting instead of sleeping for 2 minutes ?

ashinde-px replied:
We will keep it as it is as we have that standard time for FA or FB volumes. we will discuss more on this and reflect the changes after that.

ashinde-px commented:
HI @jainsmit I have addressed review comments as discussed, here is Jenkins job which is successfully run. https://jenkins.pwx.dev.purestorage.com/job/Pure-FAFB/job/Tests/view/All/job/Raw-FADA-u20/view/RawBlock-Pure-FADA/

jainsmit commented on drivers/scheduler/k8s/k8s.go:
```diff
@@ -6371,15 +6376,47 @@ func (k *K8s) CSISnapshotTest(ctx *scheduler.Context, request scheduler.CSISnaps
-		if err != nil {
-			return fmt.Errorf("failed to validate restored PVC content: %s ", err)
+		if *pvcObj.Spec.VolumeMode == corev1.PersistentVolumeBlock {
+			data := "this is pure volume rawblock test data"
```
Please add separate method to writing data into raw block. Please move code there

jainsmit commented on drivers/scheduler/k8s/k8s.go:
```diff
@@ -6412,14 +6449,53 @@ func (k *K8s) CSICloneTest(ctx *scheduler.Context, request scheduler.CSICloneReq
-		err = k.cloneAndVerify(size, data, pod.GetNamespace(), storageClassName, fmt.Sprint(request.RestoredPVCName, i), request.OriginalPVCName)
-		if err != nil {
-			return fmt.Errorf("failed to validate cloned PVC content: %s ", err)
+		if *pvcObj.Spec.VolumeMode == corev1.PersistentVolumeBlock {
```
Same, add separate method to write data into raw block mount.

ashinde-px replied:
Done

jainsmit commented on drivers/scheduler/k8s/k8s.go:
```diff
@@ -6539,15 +6622,28 @@ func (k *K8s) snapshotAndVerify(size resource.Quantity, data, snapName, namespac
-	if !strings.Contains(string(fileContent), data) {
-		return fmt.Errorf("restored volume does NOT contain data from original volume: expected to contain '%s', got '%s'", data, string(fileContent))
+	if *restoredPVC.Spec.VolumeMode == corev1.PersistentVolumeBlock {
+		ddCmd := fmt.Sprintf("dd if=%s status=none bs=38 count=1 skip=0", mountPath)
```
this code look almost same as code from line 6691 to 6710. can you please move common code here into separate method

ashinde-px replied:
Done

jainsmit commented on drivers/scheduler/k8s/k8s.go:
```diff
@@ -6590,15 +6686,28 @@ func (k *K8s) cloneAndVerify(size resource.Quantity, data, namespace, storageCla
-	}
-	if !strings.Contains(string(fileContent), data) {
-		return fmt.Errorf("cloned volume does NOT contain data from original volume: expected to contain '%s', got '%s'", data, string(fileContent))
+	if *clonedPVC.Spec.VolumeMode == corev1.PersistentVolumeBlock {
```
please incorporate with above comment

ashinde-px replied:
Done

jainsmit reviewed (approved):
LGTM !!

ashinde-px commented:
Jenkins job passed : https://jenkins.pwx.dev.purestorage.com/job/Pure-FAFB/job/Tests/view/All/job/Raw-FADA-u20/view/RawBlock-Pure-FADA/job/PureCRUDWithSDK/
